common/src/primitives/net.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use serialization::{Decode, Encode};
+
+const IPV4_SIZE: usize = 4;
+const IPV6_SIZE: usize = 16;
+
+/// IPv4 address.
+#[derive(Debug, Clone, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct Ip4 {
+    pub inner: std::net::Ipv4Addr,
+}
+
+/// IPv6 address.
+#[derive(Debug, Clone, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct Ip6 {
+    pub inner: std::net::Ipv6Addr,
+}
+
+impl Encode for Ip4 {
+    fn size_hint(&self) -> usize {
+        IPV4_SIZE
+    }
+
+    fn encode_to<T: serialization::Output + ?Sized>(&self, dest: &mut T) {
+        dest.write(&self.inner.octets());
+    }
+
+    fn encoded_size(&self) -> usize {
+        IPV4_SIZE
+    }
+}
+
+impl Decode for Ip4 {
+    fn encoded_fixed_size() -> Option<usize> {
+        Some(IPV4_SIZE)
+    }
+
+    fn decode<I: serialization::Input>(input: &mut I) -> Result<Self, serialization::Error> {
+        let mut octets = [0u8; IPV4_SIZE];
+        input.read(&mut octets)?;
+        Ok(Self {
+            inner: octets.into(),
+        })
+    }
+}
+
+impl Encode for Ip6 {
+    fn size_hint(&self) -> usize {
+        IPV6_SIZE
+    }
+
+    fn encode_to<T: serialization::Output + ?Sized>(&self, dest: &mut T) {
+        dest.write(&self.inner.octets());
+    }
+
+    fn encoded_size(&self) -> usize {
+        IPV6_SIZE
+    }
+}
+
+impl Decode for Ip6 {
+    fn encoded_fixed_size() -> Option<usize> {
+        Some(IPV6_SIZE)
+    }
+
+    fn decode<I: serialization::Input>(input: &mut I) -> Result<Self, serialization::Error> {
+        let mut octets = [0u8; IPV6_SIZE];
+        input.read(&mut octets)?;
+        Ok(Self {
+            inner: octets.into(),
+        })
+    }
+}
+
+impl From<std::net::Ipv4Addr> for Ip4 {
```

Comment:
You probably also need the conversion the other way. It's 1:1.

Reply:
Done.